lib/puppet/provider/vcsrepo/git.rb

```diff
@@ -731,16 +731,29 @@ def git_ssh_with_identity_ssh_file(*args)
 
   # Execute git with the given args, running it as the user specified.
   def exec_git(*args)
-    exec_args = {
-      failonfail: true,
-      combine: true,
-      custom_environment: { 'HOME' => Etc.getpwuid(Process.uid).dir }
-    }
-
-    if @resource.value(:user) && @resource.value(:user) != Facter['id'].value
-      exec_args[:custom_environment] = { 'HOME' => Etc.getpwnam(@resource.value(:user)).dir }
-      exec_args[:uid] = @resource.value(:user)
+    if Etc.uname[:sysname] == 'Windows_NT'
+      exec_args = {
+        failonfail: true,
+        combine: true,
+        custom_environment: { 'HOME' => Dir.home }
+      }
+      if @resource.value(:user) && @resource.value(:user) != Facter['id'].value
+        exec_args[:custom_environment] = { 'HOME' => Dir.home(@resource.value(:user)) }
+        exec_args[:uid] = @resource.value(:user)
+      end
```

Comment:
Dir.home seems portable, so using it instead of Etc.getpwnam should do the trick for all platform without the need for a code path different for Windows.

Reply:
Ah, only now I see that #553 was about fixing this now working in all circumstances.

Testing with Ruby 3.2, I can't reproduce the described issue:

```sh-session
romain@zappy /home/romain % env -i /usr/local/bin/irb
irb(main):001> ENV
=> {}
irb(main):002> Dir.home
=> "/home/romain"
irb(main):003> 
```

Reply:
So should #553 be reverted ?

Reply:
Did more testing running irb with a clean environment:

| Ruby version | `Dir.home` | Comment |
|----|---|----|
| 3.2.3 | "/home/romain" | Puppet 8.5.1 Ruby
| 3.1.2 | "/home/romain" | Debian 12 Ruby
| 2.7.6 | "/home/romain" | Puppet 7.17.0 Ruby
| 2.3.1 | ArgumentError: couldn't find HOME environment -- expanding `~' | Ubuntu 16.04 Ruby |
| 2.0.0 | ArgumentError: couldn't find HOME environment -- expanding `~' | CentOS 7 Ruby


So the version of Ruby shipped with both Puppet 7 and Puppet 8 does not exhibit this problem anymore.  I think the best is just to revert this commit @elfranne .

Reply:
I ll clean it up.

Reply:
> I ll clean it up.

Hehe :smile: 

I think the simler for you is basically:

```sh-session
$ git reset --hard origin/main
$ git revert 28c8a65f7cc09ccb647b3c93f80f805b72cd8a96
$ git push -f
```